### PR TITLE
ENH:  Add adaptive denoising module.

### PR DESCRIPTION
**Pull request by ntustison**
## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [x] Added test (or behavior not changed)
- [x] Updated API documentation (or API not changed)
- [x] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [x] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

File: Modules/Remote/AdaptiveDenoising.remote.cmake [hunk 1]
```diff
@@ -1,5 +1,5 @@
 itk_fetch_module(AdaptiveDenoising
-  "A spatially adaptive denoising image filter."
+  "A spatially adaptive denoising image filter using non-local means."
   GIT_REPOSITORY http://github.com/ntustison/ITKAdaptiveDenoising.git
```
Comment (dzenanz) [inline Modules/Remote/AdaptiveDenoising.remote.cmake:3]: Can `.DS_Store` file be removed, and preferably added to `.gitignore`?

**Review by dzenanz — CHANGES_REQUESTED** [review]
Looks like an important contribution!

[Compliance level](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Remote/README.md) should be assessed. [Example](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Remote/BoneMorphometry.remote.cmake).

File: Modules/Remote/AdaptiveDenoising.remote.cmake (hunk 1 above)
Comment (ntustison) [reply Modules/Remote/AdaptiveDenoising.remote.cmake:3]: Yep, sounds good.  Just an oversight on my part.

**Comment by ntustison** [comment]
Still a couple issues:

* [This test](https://github.com/ntustison/ITKAdaptiveDenoising/blob/master/test/itkAdaptiveNonLocalMeansDenoisingImageFilterTest.cxx#L111) is commented out because it fails [here](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/TestKernel/include/itkTestingMacros.h#L85-L86) and I don't know why.

* I thought I saw somewhere where specifying the image dimension in a class no longer follows [this style](https://github.com/ntustison/ITKAdaptiveDenoising/blob/master/include/itkNonLocalPatchBasedImageFilter.h#L52-L53)  but I can't remember where I saw this or what the alternative syntax should be.

**Comment by dzenanz** [comment]
> A lingering issue:
> 
> * I thought I saw somewhere where specifying the image dimension in a class no longer follows [this style](https://github.com/ntustison/ITKAdaptiveDenoising/blob/master/include/itkNonLocalPatchBasedImageFilter.h#L52-L53)  but I can't remember where I saw this or what the alternative syntax should be.

[That macro](https://github.com/InsightSoftwareConsortium/ITK/blob/3ecf6c87d18d51b124589d5a74389244d692109c/Modules/Core/Common/include/itkMacro.h#L762) was modified with ITKv5, but I think it is still a good choice.

It looks like you fixed the superclass test failure?

File: Modules/Remote/AdaptiveDenoising.remote.cmake [hunk 2]
```diff
@@ -1,5 +1,60 @@
...
+#--            - [ ] Linux with gcc compiler
+#--   - [ ] Active developer community dedicated to maintaining code-base
+#--   - [ ] 75% code coverage demonstrated for testing suite
+#--   - [ ] Continuous integration testing performed
+#--   - [ ] All requirements of Levels 3,2,1
+#--
+#-- ## Compliance Level 3 star (Quality beta code)
+#--   - [ ] API | executable interface is considered mostly stable and feature complete
+#--   - [ ] 10% C0-code coverage demonstrated for testing suite
+#--   - [ ] Some tests exist and pass on at least some platform
```
Comment (dzenanz) [inline Modules/Remote/AdaptiveDenoising.remote.cmake:27]: This form is supposed to be filled out.

File: Modules/Remote/AdaptiveDenoising.remote.cmake [hunk 3]
```diff
@@ -1,5 +1,60 @@
...
+  Tustison N., Avants B., Wang H., Xie L., Coupe P., Yushkevich P., Manjon J.
+  Insight Journal.
+  http://insight-journal.org/browse/publication/982.
+  Two Luis Miguel fans walk into a bar in Nagoya ---> (yada, yada, yada)
+  ---> an ITK-implementation of a popular patch-based denoising filter.
+  Tustison N., Manjon J.V.
+  Insight Journal.
+  http://insight-journal.org/browse/publication/979.
+  "
+  MODULE_COMPLIANCE_LEVEL
```
Comment (dzenanz) [inline Modules/Remote/AdaptiveDenoising.remote.cmake:57]: Some number should be given here, e.g. `MODULE_COMPLIANCE_LEVEL 3`.

**Review by dzenanz — APPROVED** [review]
Great!

File: Modules/Remote/AdaptiveDenoising.remote.cmake (hunk 3 above)
Comment (ntustison) [reply Modules/Remote/AdaptiveDenoising.remote.cmake]: Yeah, sorry about that.  I just assumed that somebody else was required to do that.

**Comment by ntustison** [comment]
Thanks @dzenanz .  I have no idea as to what I need to do next but I'll ask over at the conversation on discourse.

**Pull request merged**